### ISSUE TITLE
Feature request: Decorators support

### ISSUE THREAD

**Issue by fannheyward** [evanw/esbuild#104]
> `error: Decorators are not supported yet`

Any plan to support decorators?

**Comment by Akimyou** [comment]
https://github.com/vuejs/vue-cli/tree/dev/packages/@vue/babel-preset-app#readme

Hope can support Decorators, then we can try it in vue project.

**Comment by evanw** [comment]
The [decorator spec](https://github.com/tc39/proposal-decorators) unfortunately still seems like it's a long way off, both based on low activity and based on how many things it looks like they still have to figure out. The proposal has drifted pretty far from what the entire rest of the JavaScript ecosystem has been doing for many years. It also appears to be incompatible with a lot of the JavaScript ecosystem (e.g. no CommonJS support). So I'm not planning on supporting decorators based on the spec, at least not any time soon.

Separate from that, decorators in TypeScript have been around for a while behind the `experimentalDecorators` flag, and have found a lot of adoption in the web development community. I think esbuild would be useful to many more people if it had support for TypeScript-style decorators. So I'm planning to implement those at some point.

**Comment by wxs77577** [comment]
nest.js is a great web framework but it's extremely slow during compilation. There are over 20 modules in my project. :(

**Comment by evanw** [comment]
An initial implementation of TypeScript decorators has been released in version 0.4.10. You can read more about this in [the release notes](https://github.com/evanw/esbuild/blob/master/CHANGELOG.md#0410). Please let me know if you encounter any issues.

**Comment by guoyunhe** [comment]
I have a JS project, not TS. When compiling decorators, it throws errors:

```
[vite:esbuild] Transform failed with 1 error:
/Users/guoyunhe/src/pages/aaa/components/bbb/index.jsx:13:0: error: Unexpected "@"
file: /Users/guoyunhe/src/pages/aaa/components/bbb/index.jsx:13:0

Unexpected "@"
12 |  
   |   ^
13 |  @connect(({ aaa, bbb }) => ({
   |  ^
14 |    globalModel,
```

**Comment by evanw** [comment]
Decorators are not a JS feature, so esbuild doesn't support them in JS. Nothing in the JS specification mentions decorators: https://262.ecma-international.org/12.0/. But they are a TS feature, so esbuild supports them in TS. You can tell esbuild to compile your JS as TS if you like: `--loader:.jsx=tsx`.

**Comment by guoyunhe** [comment]
> Decorators are not a JS feature, so esbuild doesn't support them in JS. Nothing in the JS specification mentions decorators: https://262.ecma-international.org/12.0/. But they are a TS feature, so esbuild supports them in TS. You can tell esbuild to compile your JS as TS if you like: `--loader:.jsx=tsx`.

Thanks. I am now moving my project to TS and it works very well.

**Comment by UndiedGamer** [comment]
Looks like [decorators](https://github.com/tc39/proposal-decorators) but without the [metadata](https://github.com/tc39/proposals/commit/38cfceead203a9ec1d9474f395e43e3cf1427991)  has moved to stage 3 and i am sure typescript will implement this in a while, what are your plans for javascript decorators support, @evanw

**Comment by evanw** [comment]
My plans are to implement it after it has shipped in a browser, in node, or in TypeScript. This is the same thing that I do for all other syntax features.

**Comment by rluvaton** [comment]
TypeScript implemented this on 5.0 beta

https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#decorators

**Comment by evanw** [comment]
I looked into this in more detail and it seems like TypeScript is only doing a partial implementation. Work on the specification is ongoing and the API is likely going to change, at which point TypeScript will also have to change their implementation too: https://github.com/tc39/proposal-decorators/issues/494. So implementing this in esbuild still seems premature (although it's closer than before!).

**Comment by jakebailey** [comment]
I believe everything's been resolved as of today:

- https://github.com/babel/proposals/issues/86#issuecomment-1414281106
- https://github.com/microsoft/TypeScript/pull/52582

**Comment by yay** [comment]
@evanw So, when can we expect this? 😁

**Comment by Meesayen** [comment]
Hi @evanw, could this ticket be re-opened now that https://github.com/tc39/proposal-decorators/issues/494 has been closed and necessary changes been merged into the spec?

**Comment by evanw** [comment]
It looks like the specification is still actively being iterated on. For example: https://github.com/tc39/proposal-decorators/issues/499. I can reopen this issue if you want, but I still don't think it's the right time for esbuild to implement this yet as the specification is still evolving.

**Comment by Amareis** [comment]
https://github.com/tc39/proposal-decorators/issues/499 is fully merged and there is no any real issues now to implement decorators.

**Comment by wycats** [comment]
At the moment, the fact that esbuild doesn't even parse `@decorator accessor field` results in Vite reporting an error when it attempts to scan files containing that syntax for dependencies.

This makes it impossible to work around the current limitation in esbuild by using Vite plugins, since Vite assumes that it esbuild will be able to parse the syntax of JS files.

And while there have been some minor changes in semantics, the syntax of decorators has remained forwards compatible since decorators were approved for Stage 3.

Is there any chance that you would be willing to implement Stage 3 syntax in advance of implementing the full compilation?

**Comment by evanw** [comment]
Yes, I'm planning to do this. That's one of the main reasons why I did the work behind #3019. Previously esbuild considered all decorators in TypeScript files to be experimental decorators regardless of TypeScript's `experimentalDecorators` setting. This was blocking esbuild's implementation of JavaScript decorators, but that has been unblocked now that esbuild has started requiring `experimentalDecorators: true` to use experimental decorators in version 0.18.0+.

**Comment by evanw** [comment]
[Version 0.18.5](https://github.com/evanw/esbuild/releases/tag/v0.18.5) of esbuild now supports parsing and printing JavaScript decorators, and parsing and transforming auto-accessors. Please try out these features and let me know if you encounter any problems.

**Comment by Ciantic** [comment]
Did something change with 0.18.5? Decorators started to appear in transpiled code at the moment like this:

```javascript
var MonitoringApp = class {
  @(observable)
  isVisiblePerson = false;
  @(observable)
  personDetectorConnected = false;
```

And causes syntax errors.

**Comment by jakebailey** [comment]
Do you have a tsconfig with experimentalDecorators set? Otherwise, you're using the new "standard" decorators which AFAIK esbuild has not yet implemented downleveling for (and therefore stays in the output).

**Comment by Ciantic** [comment]
@jakebailey thank you!

I had to add 

1. tsconfig.json with experimentalDecorators (previously it was fine without it)
2. in esbuild config a new line `tsconfig: "tsconfig.json",`

I was using Deno to call esbuild, it worked fine a month ago but 0.18.5 indeed disabled the default setting.

**Comment by jakebailey** [comment]
Note that you can also provide `tsconfigRaw` rather than creating an all new tsconfig, though I'm not quite shure what TS setup you'd have without a `tsconfig` at all. I guess if you're trying to bundle Deno code but that's a few too many technologies for me to reason about in my head 😄

**Comment by evanw** [comment]
The [link you pasted](https://esbuild.github.io/content-types/#no-type-system) already has the answer:

> The [`emitDecoratorMetadata`](https://www.typescriptlang.org/tsconfig#emitDecoratorMetadata) TypeScript configuration option is not supported. This feature passes a JavaScript representation of the corresponding TypeScript type to the attached decorator function. Since esbuild does not replicate TypeScript's type system, it does not have enough information to implement this feature.

**Comment by MachineMakesNoise** [comment]
It would be really nice for esbuild to support typescript's new decorators (the proposal one) without the experimentalDecorators flag.

I ported my project to vite/esbuild from tsc and came across annoying problem that made me bang my head to a wall for too long :

Typescript (tsc 5.x) supports the class decorator syntax like this :
```typescript
export @decorator class Test {}
```
But with esbuild this leads to the dreaded 'Unexpected "@"...' error.
This can be fixed by moving the "export" to separate line :
```typescript
@decorator class Test {}
export { Test }
```

Hope this helps to protect someone else's head.

**Comment by justinfagnani** [comment]
Standard decorators are now implemented in TypeScript and Babel, and at least Firefox is also implementing now. I think Chrome is at least implementing the `accessor` keyword.

**Comment by iyobo** [comment]
So I guess no esbuild for any serious typescript development, given that all the major frameworks for typescript use decorators.

**Comment by iyobo** [comment]
@shigma what sort of decorator is the RIGHT decorator?

**Comment by Shakeskeyboarde** [comment]
> @shigma what sort of decorator is the RIGHT decorator? 

Legacy vs experimental decorators: https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#decorators#differences-with-experimental-legacy-decorators

**Comment by iyobo** [comment]
And your claim is that esbuild supports one over the other?

**Comment by shigma** [comment]
Of cource esbuild supports the decorator according to the ECMAScript specification.

**Comment by itpropro** [comment]
> > most major frameworks for typescript use decorators
> 
> @iyobo Those frameworks at least did not use the RIGHT decorators in a long term. I don't think they represent the whole TypeScript ecosystem.
> 
> I program in TypeScript, not DecoratorScript.

This is quite misleading as there were no other decorators available when most of the frameworks started integrating them. Decorators are an essential feature for many use cases and esbuild should support all the "legacy" versions of decorators supported by the official TypeScript config, namely "emitDecoratorMetadata" and "experimentalDecorators". There is no current replacement for the functionalities these provide, so there is in fact no "Right" decorator, as the experimental ones are the only ones supporting most use cases and are already used by many major TS frameworks for a reason.

**Comment by iyobo** [comment]
@shigma You are clearly entitled to your opinion. No argument needed. The statement stands true more for backend typescript development.

**Comment by evanw** [comment]
I have an update: I'm starting to seriously investigate implementing JavaScript decorators in esbuild (not TypeScript experimental decorators, which have already been implemented in esbuild for a long time). My first step was to write a lot of tests: https://github.com/evanw/decorator-tests. I now understand the specification for decorators much better.

As far as I understand, the main existing implementations of this WIP specification are TypeScript and Babel. Both of them seem to have some subtle bugs and/or disagreements with how this feature is supposed to behave. The problems with these tools that I have found so far are documented by https://github.com/evanw/decorator-tests.

**Comment by justinfagnani** [comment]
That test suite is awesome @evanw!

This is a bit extra, but on our team we've seen a very large emitted code size increase from standard decorators over experimental TypeScript decorators. Babel isn't better here, and I think it's mostly irreducible due to `accessor` downleveling, needing fresh context objects for each application, per-decorated-member lists of initializers, etc.

So our team is also working on a custom TypeScript transform to compiler our decorators to their non-generic non-decorator equivalents, i.e. generate custom accessors directly.

We can of course do that for esbuild too, but it does require traversing every file that might contain our decorators. It would be very nice if we could somehow filter on our decorator usage, or write some specific decorator transform function.

**Comment by evanw** [comment]
Thanks for the heads up. I am sensitive to generated code size too, and I will try to do my best to minimize the code size overhead as I implement this. I have already seen the thread over at https://github.com/microsoft/TypeScript/issues/55688. It's unfortunate that the design for this feature isn't as compact as the previous design due to all of the extra features. Part of the problem also seems to be that there is no way to syntactically opt-of of some of the features that your decorators may not even use (e.g. to skip `access` and/or `addInitializer`).

However, I'm planning to implement the specification as it is (or at least as it ends up being) without deviating from it. I think it's better for the ecosystem to coalesce around the same standard behavior instead of allowing a high level of customization. So while esbuild will have two decorator implementations (one for JavaScript decorators and one for TypeScript experimental decorators), I'm not currently planning to add the ability to have additional custom decorator implementations within esbuild itself.

**Comment by evanw** [comment]
The behavior of JavaScript decorators continues to be adjusted. I just became aware of this additional (potential) change: https://github.com/pzuraq/ecma262/pull/12. Tracking it here as I will need to take it into consideration. See also https://github.com/microsoft/TypeScript/issues/56606.

**Comment by ralyodio** [comment]
any update on this?

**Comment by emilio-toledo** [comment]
Same as @ralyodio, this would be extremely helpful when using [vitest](https://vitest.dev/) in [nestjs](https://nestjs.com/)

**Comment by what1s1ove** [comment]
> Same as @ralyodio, this would be extremely helpful when using [vitest](https://vitest.dev/) in [nestjs](https://nestjs.com/)

Hey @emilio-toledo !

ECMAScript decorators (which eabuild will support) and NestJS decorators (old typescript) are totally different. You cannot migrate from one to another.

**Comment by emilio-toledo** [comment]
@what1s1ove 

Hey thanks for the quick response. Interesting, I am able to run it using SWC, maybe they support both? Either way I am running vitest with a nestjs project using unplugin-swc so I was thinking that if esbuild which powers vite/vitest supported typescript decorators I would be able to test my nestjs modules with vitest natively. Thanks for the info.

**Comment by what1s1ove** [comment]
> @what1s1ove 
> 
> Hey thanks for the quick response. Interesting, I am able to run it using SWC, maybe they support both? Either way I am running vitest with a nestjs project using unplugin-swc so I was thinking that if esbuild which powers vite/vitest supported typescript decorators I would be able to test my nestjs modules with vitest natively. Thanks for the info.

Ahh sorry, I do not know about all vite's plugins. 
But if you want to know more about ECMAScript decorators, you can take a look at one of my article about it: 

https://dev.to/what1s1ove/ecmascript-decorators-the-ones-that-are-real-g96

**Comment by Obiwarn** [comment]
> The behavior of JavaScript decorators continues to be adjusted. I just became aware of this additional (potential) change: [pzuraq/ecma262#12](https://github.com/pzuraq/ecma262/pull/12). Tracking it here as I will need to take it into consideration. See also [microsoft/TypeScript#56606](https://github.com/microsoft/TypeScript/issues/56606).

The issues linked by @evanw seem to be fixed in the meantime. Is there anything else holding back an implementation?

**Comment by ssalbdivad** [comment]
@evanw Any updates on this? TS has been moving heavily away from legacy decorators in favor of the standard.

`tsx` has been so great in general but the lack of decorator support alone is enough that I'll probably start investigating other options soon.

**Comment by cayter** [comment]
> @evanw Any updates on this? TS has been moving heavily away from legacy decorators in favor of the standard.
> 
> I've been using esbuild via`tsx` which has been so great in general- zero issues with source maps or transpilation. That said, the lack of decorator support alone is enough that I'd probably end up investigating other options soon if there's not a solution, which would honestly suck.

Our saviour has been working hard in the past 2 weeks. Let's give him some space. 

https://github.com/evanw/decorator-tests

**Comment by ssalbdivad** [comment]
@cayter Amazing, thanks for the update!

It's a fine line between highlighting the relative importance of an issue (helpful) and adding undue stress on an issue the maintainer has already prioritized or is working on (unhelpful).

I'm very sorry if my comment came across as the former. @evanw I deeply appreciate your work and just seeing this is actively being worked on alleviates my primary concern.

**Comment by alex3683** [comment]
@ssalbdivad While waiting on this to be finished I'm successfully using this vite plugin as workaround to precompile typescript: https://github.com/herberttn/vite-plugin-typescript-transform
I don't know whether you are using plain esbuild or as part of vite, so it may help or not, but maybe it helps others.

**Comment by evanw** [comment]
I'm still making progress on this. However, progress is slow because decorators are very complicated. Some things I did recently:

* Wrote a bunch more tests: https://github.com/evanw/decorator-tests (note: this requires time-consuming original research as AFAIK decorators is [mostly without tests](https://github.com/tc39/test262/issues/4042))
* Filed another bug with Babel: https://github.com/babel/babel/issues/16473
* Filed more bugs with TypeScript: https://github.com/microsoft/TypeScript/issues/58436
* Asked for clarification about the spec: https://github.com/tc39/proposal-decorators/issues/529

**Comment by evanw** [comment]
Ok [version 0.21.0](https://github.com/evanw/esbuild/releases/v0.21.0) is out and now supports the JavaScript decorators proposal. Please try it out and let me know what you think.

@justinfagnani Code size definitely isn't optimal but it shouldn't be that bad I think. I made a trade-off that generates accessor methods at run-time instead of compile-time. It would be interesting to see where esbuild's implementation ends up in your size comparison (is it public somewhere perhaps?).

**Comment by cayter** [comment]
> Ok [version 0.21.0](https://github.com/evanw/esbuild/releases/v0.21.0) is out and now supports the JavaScript decorators proposal. Please try it out and let me know what you think.
> 
> @justinfagnani Code size definitely isn't optimal but it shouldn't be that bad I think. I made a trade-off that generates accessor methods at run-time instead of compile-time. It would be interesting to see where esbuild's implementation ends up in your size comparison (is it public somewhere perhaps?).

This is crazy. I thought this would take another few months to ship. Thanks a lot!

**Comment by evanw** [comment]
> Note that I couldn't get the example in the release working

Here's an example of it working: [link](https://esbuild.github.io/try/#dAAwLjIxLjAALS10YXJnZXQ9ZXMyMDIyAGNvbnN0IGxvZyA9IChmbiwgY29udGV4dCkgPT4gZnVuY3Rpb24oKSB7CiAgY29uc29sZS5sb2coYGJlZm9yZSAke2NvbnRleHQubmFtZX1gKQogIGNvbnN0IGl0ID0gZm4uYXBwbHkodGhpcywgYXJndW1lbnRzKQogIGNvbnNvbGUubG9nKGBhZnRlciAke2NvbnRleHQubmFtZX1gKQogIHJldHVybiBpdAp9CgpjbGFzcyBGb28gewogIEBsb2cgc3RhdGljIGZvbygpIHsKICAgIGNvbnNvbGUubG9nKCdpbiBmb28nKQogIH0KfQoKLy8gTG9ncyAiYmVmb3JlIGZvbyIsICJpbiBmb28iLCAiYWZ0ZXIgZm9vIgpGb28uZm9vKCk). That logs this when run:

```
before foo
in foo
after foo
```

**Comment by cayter** [comment]
> > Note that I couldn't get the example in the release working
> 
> Here's an example of it working: [link](https://esbuild.github.io/try/#dAAwLjIxLjAALS10YXJnZXQ9ZXMyMDIyAGNvbnN0IGxvZyA9IChmbiwgY29udGV4dCkgPT4gZnVuY3Rpb24oKSB7CiAgY29uc29sZS5sb2coYGJlZm9yZSAke2NvbnRleHQubmFtZX1gKQogIGNvbnN0IGl0ID0gZm4uYXBwbHkodGhpcywgYXJndW1lbnRzKQogIGNvbnNvbGUubG9nKGBhZnRlciAke2NvbnRleHQubmFtZX1gKQogIHJldHVybiBpdAp9CgpjbGFzcyBGb28gewogIEBsb2cgc3RhdGljIGZvbygpIHsKICAgIGNvbnNvbGUubG9nKCdpbiBmb28nKQogIH0KfQoKLy8gTG9ncyAiYmVmb3JlIGZvbyIsICJpbiBmb28iLCAiYWZ0ZXIgZm9vIgpGb28uZm9vKCk). That logs this when run:
> 
> ```
> before foo
> in foo
> after foo
> ```

Ah I didn't specify target. Thanks man!

**Comment by Obiwarn** [comment]
![image](https://github.com/evanw/esbuild/assets/9539797/696050ba-302c-4c38-b0ea-b1bdb80ce378)

The grind was real.

![image](https://github.com/evanw/esbuild/assets/9539797/1f39e9b0-0330-4fc1-956d-fc11542b3dfd)

Thank you so much @evanw!

I will test this thoroughly with our internal ORM.

**Comment by alex3683** [comment]
I tried it in our product but got an error with MobX and thus had to return to the transformer workaround. I then tried make a small reproduction by using your online esbuild tranformer and putting it in a jsfiddle , but there it runs without problems. I'll just post the error and what I found out so far what goes wrong, but I guess it's not easy to debug.

MobX just throws this: `Uncaught TypeError: Cannot convert undefined or null to object` (but only for us, not in the fiddle). This is in an initializer of an accessor, where `this` is undefined. I tried to trace bug how esbuild transforms the code, but I didn't really grog the flags parameter that goes in. The only thing I understand is, that when its even, `this` is undefined, and when it's odd, its set. MobX however is called with an even flag. I tried to find out what drives the decision within esbuild, but it was too hard for me in a few hours and reverted. I can only offer you to have remote look at my code via teams or something in case you are interested. Or make adjustments that may lead to the cause of this. I guess when this is solved, I'd give you a fully positive feedback :-D

This was my reproduction attempt:
- https://esbuild.github.io/try/#dAAwLjIxLjEAewogIHRhcmdldDogWydlczIwMTUnXSwKICBsb2FkZXI6ICd0cycsCiAgdHNjb25maWdSYXc6IHsKICAgIGNvbXBpbGVyT3B0aW9uczogewogICAgICBhbGxvd0pzOiBmYWxzZSwKICAgICAgYWxsb3dTeW50aGV0aWNEZWZhdWx0SW1wb3J0czogdHJ1ZSwKICAgICAgY29tcG9zaXRlOiB0cnVlLAogICAgICBkZWNsYXJhdGlvbjogdHJ1ZSwKICAgICAgZGVjbGFyYXRpb25NYXA6IHRydWUsCiAgICAgIGVzTW9kdWxlSW50ZXJvcDogdHJ1ZSwKICAgICAgZm9yY2VDb25zaXN0ZW50Q2FzaW5nSW5GaWxlTmFtZXM6IHRydWUsCiAgICAgIGlzb2xhdGVkTW9kdWxlczogdHJ1ZSwKICAgICAganN4OiAncmVhY3QtanN4JywKICAgICAgbGliOiBbJ2RvbScsICdkb20uaXRlcmFibGUnLCAnZXNuZXh0J10sCiAgICAgIG1vZHVsZTogJ2VzbmV4dCcsCiAgICAgIG1vZHVsZVJlc29sdXRpb246ICdub2RlJywKICAgICAgbm9FbWl0OiB0cnVlLAogICAgICBub0ltcGxpY2l0T3ZlcnJpZGU6IGZhbHNlLAogICAgICBub0ZhbGx0aHJvdWdoQ2FzZXNJblN3aXRjaDogZmFsc2UsCiAgICAgIG5vVW51c2VkTG9jYWxzOiBmYWxzZSwKICAgICAgcmVzb2x2ZUpzb25Nb2R1bGU6IHRydWUsCiAgICAgIHNraXBMaWJDaGVjazogdHJ1ZSwKICAgICAgc3RyaWN0OiBmYWxzZSwKICAgICAgdGFyZ2V0OiAnZXMyMDE1JywKICAgIH0sCiAgfSwKfQBjb25zdCB7b2JzZXJ2YWJsZSwgcmVhY3Rpb24gfSA9IG1vYng7CgpjbGFzcyBGb28gewogIEBvYnNlcnZhYmxlCiAgcHVibGljIGFjY2Vzc29yIHRleHQ6IHN0cmluZyA9ICJIZWxsbyBBY2Nlc3NvciI7Cn0KCmNvbnN0IGZvbyA9IG5ldyBGb28oKTsKcmVhY3Rpb24oKCkgPT4gZm9vLnRleHQsIHRleHQgPT4gewogY29uc29sZS5sb2coIlRleHQgY2hhbmdlZCB0bzogIiArIHRleHQpOwp9KTsKY29uc29sZS5sb2coZm9vLnRleHQpOwpmb28udGV4dCA9ICJHb29kYnllIjs
- https://jsfiddle.net/7yvw5fz3/

**Comment by hrishikeshs** [comment]
Wow! I was literally trying to write a decorator yesterday to add logging to datadog to our common logger utility. Can't wait to try this out 👍  thank you so much!

**Comment by marvinruder** [comment]
> What do people think? Does it work with existing code that uses the JavaScript decorators proposal? Any feedback?

Works great! I was able to migrate my decorators from the legacy Typescript implementation to the TC39 proposal without problems. The only issue I encountered was that class field decorators have no access to the class itself, but this is a limitation of the specification and [already discussed there](https://github.com/tc39/proposal-decorators/issues/466).

**Comment by justinfagnani** [comment]
@evanw I tried the Lit decorators, but it looks like metadata isn't implemented yet, which is a blocker for us. I'll be able to test again when metadata support is implemented.

**Comment by evanw** [comment]
Initial support for decorator metadata has been released in [version 0.21.3](https://github.com/evanw/esbuild/releases/tag/v0.21.3).

**Comment by b3nten** [comment]
Evan you rock!

**Comment by justinfagnani** [comment]
Awesome @evanw!

I ran all of Lit's decorator tests after an esbuild compile and they all pass! I'll [check those tests in](https://github.com/lit/lit/pull/4648) so if you ever want to check changes against them you can do:

```sh
git clone https://github.com/lit/lit.git
cd lit
npm ci
npm link esbuild
npm run test:dev -w @lit/reactive-element
```

**Comment by mprinc** [comment]
> Initial support for decorator metadata has been released in [version 0.21.3](https://github.com/evanw/esbuild/releases/tag/v0.21.3).

Hi @evanw In [the esbuild documentation](The [emitDecoratorMetadata](https://www.typescriptlang.org/tsconfig#emitDecoratorMetadata) TypeScript configuration option is not supported) you still have the statement that: The [`emitDecoratorMetadata`](https://www.typescriptlang.org/tsconfig#emitDecoratorMetadata) TypeScript configuration option is not supported

1. Is that obsolated now
2. Does anyone know and/or can confirm that we can use esbuild now for nest.js projects?
3. Does TSX work well now in regard to all the 3 types of decorators that are now implemented in `esbuild`?

**Comment by evanw** [comment]
As far as I know the documentation about this is up to date. There are only two types of decorators supported by esbuild: legacy TypeScript decorators and standard JavaScript decorators. Furthermore all TypeScript features that require a type system to function (including emitDecoratorMetadata) are not supported by esbuild, as esbuild doesn’t have a type system. Code that uses TypeScript legacy decorators without emitDecoratorMetadata should run fine under esbuild.

**Comment by Arlen22** [comment]
> As far as I know the documentation about this is up to date. There are only two types of decorators supported by esbuild: legacy TypeScript decorators and standard JavaScript decorators. Furthermore all TypeScript features that require a type system to function (including emitDecoratorMetadata) are not supported by esbuild, as esbuild doesn’t have a type system. Code that uses TypeScript legacy decorators without emitDecoratorMetadata should run fine under esbuild.

https://esbuild.github.io/content-types/#tsconfig-json 

This section says that decorators are not transformed when "experimentalDecorators" is disabled.

**Comment by danielrentz** [comment]
@Arlen22 I wrote https://github.com/esbuild/esbuild.github.io/issues/54
